pkg/stub/handler.go

```diff
@@ -2,96 +2,436 @@ package stub
+			return h.processError(operatorCfg, v1alpha1.SecretUpdateFailed, err, "failed to delete before create dockerconfig secret the openshift namespace: %v")
+		}
+		logrus.Printf("registry dockerconfig secret %s was deleted", secret.Name)
+		newStatus = corev1.ConditionFalse
+	} else {
+		secretToCreate := corev1.Secret{}
+		secret.DeepCopyInto(&secretToCreate)
+		secretToCreate.Namespace = ""
+		secretToCreate.ResourceVersion = ""
+
+		s, err := h.secretclientwrapper.Get("openshift", secretToCreate.Name)
+		if err != nil && !kerrors.IsNotFound(err) {
+			return h.processError(operatorCfg, v1alpha1.SecretUpdateFailed, err, "failed to get registry dockerconfig secret in openshift namespace : %v")
+		}
 		if err != nil {
-			logrus.Errorf("failed to get rest config : %v", err)
-			return err
+			s = nil
+		}
+		if s != nil {
+			_, err = h.secretclientwrapper.Update("openshift", &secretToCreate)
+		} else {
+			_, err = h.secretclientwrapper.Create("openshift", &secretToCreate)
 		}
-		tempclient, err := getTemplateClient(restconfig)
 		if err != nil {
-			logrus.Errorf("failed to get template client : %v", err)
-			return err
+			return h.processError(operatorCfg, v1alpha1.SecretUpdateFailed, err, "failed to create/update registry dockerconfig secret in openshif namespace : %v")
+		}
+		logrus.Printf("created/updated dockerconfig secret %s in openshift namespace", operatorCfg.Spec.RegistryCredentials.Name)
+		newStatus = corev1.ConditionTrue
+	}
+
+	now := kapis.Now()
+	status := operatorCfg.Condition(v1alpha1.ImportCredentialsExists)
+	status.LastUpdateTime = now
+	if status.Status != newStatus {
+		status.Status = newStatus
+		status.LastTransitionTime = now
+	}
+	operatorCfg.ConditionUpdate(status)
+	err := h.sdkwrapper.Update(operatorCfg)
+	if err != nil {
+		if !kerrors.IsConflict(err) {
+			return h.processError(operatorCfg, v1alpha1.SamplesUpdateFailed, err, "failed adding success condition to config: %v")
+		}
+		logrus.Println("got conflict error on success status update, just going to forgo this time vs. retry")
+		// just return the raw error and initiate the sdk's requeue/ratelimiter setup
+		return err
+	}
+
+	return nil
+}
+
+func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
+	switch event.Object.(type) {
+	case *corev1.Secret:
+		dockercfgSecret, _ := event.Object.(*corev1.Secret)
+
+		if h.operatorCfg == nil {
+			dockercfgSecret.DeepCopyInto(&h.registrySecret)
+			return nil
+		}
+
+		if h.operatorCfg.Spec.RegistryCredentials.Name == dockercfgSecret.Name {
+			err := h.manageDockerCfgSecret(event.Deleted, h.operatorCfg, dockercfgSecret)
+			if err != nil {
+				return err
+			}
+		}
+
+	case *v1alpha1.SamplesResource:
+		newStatus := corev1.ConditionTrue
+		var opcfg *v1alpha1.SamplesResource
+		if event.Deleted {
+			h.operatorCfg = nil
+			// "4a" in the "startup" workflow, just create default
+			// resource and set up that way
+			opcfg = &v1alpha1.SamplesResource{}
```

Comment:
Where do you actually create this resource?
Also we should be creating the resource w/ the default values set, rather than determing the defaults in code.  That way someone who looks at the resource instance can see what it is going to create.

Reply:
basically in case of event.Deleted (or on initial startup if no SamplesResource exists) you should instantiate a new SamplesResource (with the defaults populated) and return.  Then you'll get a "SamplesResource created" event which you'll process as normal.

Reply:
OK, in the case of event.Deleted, seems straight forward.

The initial start up is trickier, as they may be creating the deployment and samples resource at the same time.  Because of that don't think you can just test for existence during the launch in main.go.

Something like a one time timer pop after a few seconds or something seems better.

Opinions?

Reply:
fair point.  in general i think our expected pattern will be:
1) create operator
2) let operator create default resource
3) user tweaks resource if needed

but i can envision someone wanting to do what you described (create the resource and operator together because they are opinionated about the resource they want to create and they don't want to mutate the default one after it is created).  (though our installer will likely never do it that way).

I think your suggestion of having the operator startup logic set a timer and wait 5-10s before creating the default resource (assuming a resource hasn't been created by then) is a decent compromise to allow for that.

Reply:
part of next push